einit/documentation/man/einit-control.8
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "EINIT-CONTROL" "8" "19 February 2007" "einit-0.15.1-svn" ""

.SH NAME
einit-control \- chat with einit using IPC commands
.SH SYNOPSIS

\fBeinit-control\fR [ \fB-s \fIcontrol-socket\fB\fR ] [ \fB-h\fR ] [ \fB-v\fR ] [ \fB--\fR ] [ \fB\fIipc-command\fB\fR\fI ...\fR ]

.SH "DESCRIPTION"
.PP
\fBeinit-control\fR is used to interface with the init-replacement \fBeinit\fR\&.
It does so by opening einit's control socket and sending/receiving text strings that are processed
by einit or its modules.
.PP
\fBeinit\fR must be running for this command to do anything useful, also the \fBipc module\fR must enabled.
.SH "OPTIONS"
.TP
\fB-s \fIcontrol-socket\fB\fR
Use \fIcontrol-socket\fR instead of /etc/einit-control.
.TP
\fB-v\fR
Display version and license strings, then exit.
.TP
\fB-h\fR
Display help screen and exit.
.TP
\fB--\fR
Terminate list of arguments to einit-control, send the rest of the arguments to einit's IPC
socket verbatim.
.TP
\fB[\fIipc-command\fB\&...]\fR
The actual commands to execute. When specifying multiple commands, they need to be delimited
using semi-colons.
.SH "IPC COMMANDS"
.PP
Of course you'll need to know about some IPC commands to do anything useful with this program.
This section is divided into subsections by the first element of the command.
.sp
.RS
.B "Note:"
The available IPC commands depend on the loaded modules (among other factors). The ones listed
here are provided using the default configuration, as of the time of this writing. Also, some
potentially hazardous and debug-only commands were left out.
.RE
.SS "RC"
commands to manipulate the current system status
.TP
\fBrc \fIservice\fB enable\fR
Bring up \fIservice\fR, possibly including others that this service requires.
.TP
\fBrc \fIservice\fB disable\fR
Bring down \fIservice\fR\&.
.TP
\fBrc \fIservice\fB reset\fR
Reset \fIservice\fR\&. Not implemented by many modules.
.TP
\fBrc \fIservice\fB reload\fR
Tell \fIservice\fR to re-load itself. Not implemented by many modules.
.TP
\fBrc \fIservice\fB zap\fR
Tell eINIT that \fIservice\fR is no longer provided.
.TP
\fBrc switch-mode \fImode\fB\fR
Tell einit to switch to \fImode\fR\&.
.SS "POWER"
manipulate system power status
.TP
\fBpower down\fR
Shut down the system: switch to the mode "power-down" and tell the kernel to turn the computer off.
.TP
\fBpower reset\fR
Reset the system: switch to the mode "power-reset" and tell the kernel to reboot.
.SS "LIST"
commands to list certain program stats
.TP
\fBlist modules\fR
Display a list of modules, including their current status, the services they provide and
require and before and after which services they request to be enabled.
.TP
\fBlist services\fR
Display a list of services that are currently available. The parameter --ansi will make this
output colourful, --only-relevant will only display services that are used in one mode or another.
.TP
\fBlist configuration\fR
Query eINIT for a list of all the configuration entries.
.TP
\fBlist configuration \fIregex\fB\fR
Query eINIT for a list of all the configuration entries where the variable name matches the regular
expression \fIregex\fR\&.
.SS "UPDATE"
commands to update the system status
.TP
\fBupdate configuration\fR
Tell einit's core and modules to re-read all configuration sources as so far.
.TP
\fBupdate configuration \fIfile\fB\fR
Tell einit's core and modules to try and use \fIfile\fR as a source for
configuration data.
.SS "EXAMINE"
examine certain program stats
.TP
\fBexamine configuration\fR
Examine the current configuration and report problems and potential errors.
(This is the same data reported by \fBeinit --wtf\fR, except that it's from the
currently running instance of einit's point of view.)
.SS "IPC COMMAND OPTIONS"
IPC commands may have flags of their own, a few common ones are:
.TP
\fB--help\fR
Tell the intepreting module(s) to display a help screen instead of running the command.
(Not really implemented.)
.TP
\fB--xml\fR
Tell the intepreting module(s) to provide feedback in XML instead of regular text.
You mileage may vary.
.TP
\fB--only-relevant\fR
Tell interpreting module(s) to only report "relevant" information. The exact interpretation
of this is up to the module and depends on the command in question, for example
"list modules --only-relevant" will list only those modules that have had a status change
happen to them.
.TP
\fB--detach\fR
Tell the processing module(s) to take the command and process it internally, allowing
einit-control to return right after querying the command. Only implemented for commands
of the "rc"-type.
.SH "USAGE"
.PP
Popular commands (the hash means "enter everything after that on a root console):
.PP

.nf
    # einit-control power down
    # einit-control update configuration
    # einit-control rc switch-mode mymode
    # einit-control rc sshd enable
   
.fi
.PP
This program is, in fact, a multi-call binary. You may symlink to /sbin/einit-control to imply
the first part of an IPC command, as is done with the "power (down|reset)" and "(e)rc" commands.
For example, the following two commands are equivalent, provided that there is a symlink linking
"power" to "einit-control":
.PP

.nf
    # einit-control power down
    # power down
   
.fi
.SH "FILES"
.TP
\fB/etc/einit-control\fR
The default control socket.
.SH "BUGS"
.PP
Please send bug reports to einit's bugtracker on sourceforge.net or add them to einit's wiki.
.PP
Remember that this is BETA software: avoid using this on mission-critical systems.
.TP 0.2i
\(bu
\fBeinit-control\fR may get "stuck" while reading from einit's control socket. If that happens to you, simply hit CTRL+C.
.SH "SEE ALSO"
.PP
einit(8), power(8), erc(8)
.PP
The official project page <URL:http://einit.sourceforge.net/> and the official wiki <URL:http://wiki.jyujin.de/>
.SH "AUTHOR"
.PP
Written by Magnus Deininger <mdeininger@jyujin.de>\&.
.SH "COPYRIGHT"
.PP
(C) 2006 Magnus Deininger, All rights reserved.
.PP
Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:
.PP
.TP 0.2i
\(bu
Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.
.TP 0.2i
\(bu
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
.TP 0.2i
\(bu
Neither the name of the project nor the names of its contributors may be
used to endorse or promote products derived from this software without
specific prior written permission.
.PP
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


einit/src/einit-control.c
/*
 *  einit-control.c
 *  einit
 *
 *  Created by Magnus Deininger on 03/05/2006.
 *  Copyright 2006, 2007 Magnus Deininger. All rights reserved.
 *
 */

/*
Copyright (c) 2006, 2007, Magnus Deininger
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
	  this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
	  this list of conditions and the following disclaimer in the documentation
	  and/or other materials provided with the distribution.
    * Neither the name of the project nor the names of its contributors may be
	  used to endorse or promote products derived from this software without
	  specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <einit/bitch.h>
#include <einit/utility.h>
#include <einit/config.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <string.h>
#include <libgen.h>
#include <errno.h>

#include <sys/ioctl.h>

int main(int, char **);
int print_usage_info ();
int ipc (char *);

char *ctrlsocket = "/etc/einit-control";

int print_usage_info () {
 if (fputs ("eINIT " EINIT_VERSION_LITERAL " Control\nCopyright (c) 2006, 2007, Magnus Deininger\nUsage:\n einit-control [-s control-socket] [-v] [-h] [function] [--] command\n [function] [-s control-socket] [-v] [-h] [--] command\n\npossible commands for function \"power\":\n down   tell einit to shut down the computer\n reset  reset/reboot the computer\n\nNOTE: calling einit-control [function] [command] is equivalent to calling [function] [command] directly.\n  (provided that the proper symlinks are in place.)\n", stderr) < 0)
  bitch2(BITCH_STDIO, "print_usage_info", 0, "fputs() failed.");
 return -1;
}

int ipc (char *cmd) {
 int sock = socket (AF_UNIX, SOCK_STREAM, 0), ret = 0;
 char buffer[1024];
 struct sockaddr_un saddr;
 int len = strlen (cmd);
 char *c;
 FILE *esocket;
 if (sock == -1)
  return bitch (BTCH_ERRNO);

 saddr.sun_family = AF_UNIX;
 strncpy (saddr.sun_path, ctrlsocket, sizeof(saddr.sun_path) - 1);

 if (connect(sock, (struct sockaddr *) &saddr, sizeof(struct sockaddr_un))) {
  close (sock);
  return bitch (BTCH_ERRNO);
 }

 len = strlen(cmd);
 c = emalloc ((len+2)*sizeof (char));
 snprintf (c, (len+2), "%s\n", cmd);

 if (!(esocket = fdopen (sock, "w+"))) {
  fputs (" >> opening socket failed.", stderr);
  bitch(BTCH_ERRNO);
  close (sock);
  return 0;
 }

 if (fputs(c, esocket) == EOF) {
  fputs (" >> sending command failed.", stderr);
  bitch(BTCH_ERRNO);
 }

 fflush (esocket);

 while ((!feof(esocket)) && fgets (buffer, 1024, esocket)) {
  if (!strcmp("IPC//processed.\n", buffer)) {
   char retval[512];
   *retval = 0;

   fgets (retval, 512, esocket);

   ret = atoi(retval);

   break;
  }

  if (fputs (buffer, stdout) < 0)
   bitch2(BITCH_STDIO, "ipc", 0, "fputs() failed.");
 }

 errno = 0;

 fclose (esocket);
 return ret;
}

int main(int argc, char **argv) {
 int i, l, ret = 0;
 char *c = emalloc (1*sizeof (char));
 char *name = estrdup ((char *)basename(argv[0]));
 char ansi = 0;
 c[0] = 0;
 if (!strcmp (name, "erc")) {
  c = (char *)erealloc (c, 3*sizeof (char));
  c = strcat (c, "rc");
 } else if (strcmp (name, "einit-control")) {
  c = (char *)erealloc (c, (1+strlen(name))*sizeof (char));
  c = strcat (c, name);
 }

 struct winsize ws;
 if (!ioctl (1, TIOCGWINSZ, &ws) || !(errno == ENOTTY)) {
  ansi = 1;
 }

 for (i = 1; i < argc; i++) {
  if (argv[i][0] == '-')
   switch (argv[i][1]) {
    case 's':
     if ((++i) < argc)
      ctrlsocket = argv[i];
     else
      return print_usage_info ();
     break;
    case 'h':
     return print_usage_info ();
     break;
    case 'v':
     if (puts("eINIT " EINIT_VERSION_LITERAL "\nCopyright (c) 2006, 2007, Magnus Deininger") < 0)
      bitch2(BITCH_STDIO, "main", 0, "puts() failed.");
     return 0;
    case '-':
     i++;
     if (i < argc) goto copy_remainder_verbatim;
     return 0;
   }
  else while (i < argc) {
   copy_remainder_verbatim:
   l = strlen(c);
   if (l) {
    c = erealloc (c, (l+2+strlen(argv[i]))*sizeof (char));
    c[l] = ' ';
    c[l+1] = 0;
   } else {
    c = erealloc (c, (1+strlen(argv[i]))*sizeof (char));
   }
   c = strcat (c, argv[i]);

   i++;
  }
 }

 if (ansi) {
  c = erealloc (c, (strlen(c)+8)*sizeof (char));
  c = strcat (c, " --ansi");
 }

 ret = ipc(c);

 bitch (BTCH_DL + BTCH_ERRNO);
 return ret;
}
